src/types/dict.rs

```diff
@@ -228,51 +228,129 @@ impl PyDict {
+        di_used: ffi::Py_ssize_t,
+        marker: PhantomData<&'py PyDict>,
+    }
+
+    impl<'py> Iterator for PyDictIterator<'py> {
+        type Item = (&'py PyAny, &'py PyAny);
+
+        /// Advances the iterator and returns the next value.
+        ///
+        /// # Panics
+        ///
+        /// If PyO3 detects that the dictionary is mutated during iteration, it will panic.
+        /// It is allowed to modify the values of the keys as you iterate over the dictionary, but only
+        /// so long as the set of keys does not change.
+        #[inline]
+        fn next(&mut self) -> Option<Self::Item> {
+            unsafe {
+                if self.di_used != (*(self.di_dict)).ma_used {
+                    self.di_used = -1;
+                    panic!("dictionary changed size during iteration");
+                };
+
+                let mut key: *mut ffi::PyObject = std::ptr::null_mut();
+                let mut value: *mut ffi::PyObject = std::ptr::null_mut();
+                if ffi::PyDict_Next(self.di_dict.cast(), &mut self.di_pos, &mut key, &mut value)
+                    != 0
+                {
+                    let py = Python::assume_gil_acquired();
+                    // PyDict_Next returns borrowed values; for safety must make them owned (see #890)
+                    Some((
+                        py.from_owned_ptr(ffi::_Py_NewRef(key)),
+                        py.from_owned_ptr(ffi::_Py_NewRef(value)),
+                    ))
+                } else {
+                    None
+                }
+            }
+        }
 
-impl<'py> Iterator for PyDictIterator<'py> {
-    type Item = (&'py PyAny, &'py PyAny);
+        #[inline]
+        fn size_hint(&self) -> (usize, Option<usize>) {
+            let len = unsafe { ffi::PyDict_Size(self.di_dict.cast()) as usize };
+            (
+                len.saturating_sub(self.di_pos as usize),
+                Some(len.saturating_sub(self.di_pos as usize)),
+            )
+        }
+    }
 
-    #[inline]
-    fn next(&mut self) -> Option<Self::Item> {
-        unsafe {
-            let mut key: *mut ffi::PyObject = std::ptr::null_mut();
-            let mut value: *mut ffi::PyObject = std::ptr::null_mut();
-            if ffi::PyDict_Next(self.dict.as_ptr(), &mut self.pos, &mut key, &mut value) != 0 {
-                let py = self.dict.py();
-                // PyDict_Next returns borrowed values; for safety must make them owned (see #890)
-                Some((
-                    py.from_owned_ptr(ffi::_Py_NewRef(key)),
-                    py.from_owned_ptr(ffi::_Py_NewRef(value)),
-                ))
-            } else {
-                None
+    impl<'a> std::iter::IntoIterator for &'a PyDict {
+        type Item = (&'a PyAny, &'a PyAny);
+        type IntoIter = PyDictIterator<'a>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            unsafe {
+                let di_dict: *mut ffi::PyDictObject = self.as_ptr().cast();
```

Comment:
I've always wondered if there should be an `as_raw() -> *mut ffi::PyDictObject` method on `PyDict` (and similar for other types)? Available only on not (limited_api / PyPy).